shared/lib-hoppy/hoppy/async_hoppy_client.py

```diff
@@ -0,0 +1,204 @@
+                                            routing_key=reply_routing_key,
+                                            reply_callback=self._on_reply)
+
+    def start(self, loop):
+        self.async_publisher.connect(loop)
+        self.async_consumer.connect(loop)
+
+    def stop(self):
+        self.async_publisher.stop()
+        self.async_consumer.stop()
+
+    async def make_request(self, request_id, body):
+        correlation_id = str(uuid.uuid4())
+        logging.info(
+            f"event=requestStarted id={request_id} "
+            f"client={self.name} "
+            f"queue={self.request_queue_properties.name} "
+            f"correlation_id={correlation_id}")
+        self.responses[correlation_id] = None
+
+        try:
+            self.async_publisher.publish_message(body,
+                                                 BasicProperties(app_id=self.app_id,
+                                                                 content_type="application/json",
+                                                                 reply_to=self.reply_queue_properties.name,
+                                                                 correlation_id=correlation_id))
+        except Exception as e:
+            logging.warning(f"event=requestError "
+                            f"client={self.name} "
+                            f"id={request_id} "
+                            f"queue={self.request_queue_properties.name} "
+                            f"correlation_id={correlation_id} "
+                            f"error='{UNEXPECTED_PUBLISH_ERROR}: {e}'")
+            raise ResponseException(message=UNEXPECTED_PUBLISH_ERROR)
+
+        wait_for_response_time = None
+        while True:
+            response = self.responses.get(correlation_id)
+            if response is not None:
+                self._terminate_correlation_id(correlation_id)
+                if isinstance(response, json.JSONDecodeError):
+                    logging.warning(f"event=requestError "
+                                    f"client={self.name} "
+                                    f"id={request_id} "
+                                    f"queue={self.request_queue_properties.name} "
+                                    f"correlation_id={correlation_id} "
+                                    f"error='{UNDECODABLE}'")
+                    raise ResponseException(message=UNDECODABLE)
+                else:
+                    logging.info(f"event=requestCompleted "
+                                 f"client={self.name} "
+                                 f"id={request_id} "
+                                 f"queue={self.request_queue_properties.name} "
+                                 f"correlation_id={correlation_id}")
+                    return response
+            else:
+                if wait_for_response_time is None:
+                    wait_for_response_time = time.time()
+                    await asyncio.sleep(0)
```

Comment:
I'm worried there might be a scary coroutine bug here -- `await asyncio.sleep(0)` only gets called the first time through the loop. In the off-chance that no response arrives before awaking, won't the method just keep spinning hard without yielding, until timeout?

Reply:
Ooo good catch. 🎯